File: python/BST.py
class ST:
    __size=0
    __root=None
    class __Node:

       def __init__(self,key,value):
           self.key=key
           self.value=value
           self.leftChild=None
           self.rightChild=None    

    def __init__(self,key,value):
        if ST.__root==None:
            ST.__root=ST.__Node(key,value)
            ST.__size+=1
    

    def size(self):
        return ST.__size

    def put(self,key,value):
        if ST.__root!=None:
           self.__add(ST.__root,key,value)
           ST.__size+=1
        else:
         ST.__root=ST.__Node(key,value)   


    def r(self):
        return ST.__root       
            


    def __add(self,root,key,value):
            cmp=root.key 
            if key>cmp:
                if root.rightChild==None:
                    root.rightChild=ST.__Node(key,value)
                    return
                else:
                   self.__add(root.rightChild,key,value)
            else:
                if root.leftChild==None:
                    root.leftChild=ST.__Node(key,value)
                    return
                else:
                   self.__add(root.leftChild,key,value)    


    def contains(self,key):
        if ST.__size==0:
             return "Tree is empty"

        return self.__search(ST.__root,key) 

    def min(self):
        if ST.__root==None:
            return "Tree is empty"
        return self.__findMin(ST.__root)

    def max(self):
        if ST.__root==None:
            return "Tree is empty"
        return self.__findMax(ST.__root)    
    
    def __findMax(self,node):
        if node.rightChild==None:
             return node.key
        return self.__findMax(node.rightChild)   

    def __findMin(self,node):
        if node.leftChild==None:
            return node.key
        return self.__findMin(node.leftChild)  
    
    def __search(self,root,key):
        if root==None:
          return "Not found"

        if key==root.key:
              return root.value

        if key<root.key:
           return self.__search(root.leftChild,key)            

        if key>root.key:
            return  self.__search(root.rightChild,key)    



# Runner code
s=ST("F",6)
s.put("E",3)
s.put("C",8)
s.put("B",0)

print(s.min())

File: java/BST/BST.java
import java.util.LinkedList;
import java.util.Queue;

class BST<K,V> {
private Node<K,V> root=null;

 BST(K key,V val){
 put(key,val);
 }


 private class Node<K,V> {
 private K key;
 private V value;
private int count;
 private Node<K, V> rightChild = null;
 private Node<K, V> leftChild = null;

  Node(K key, V value) {
   this.key = key;
   this.value = value;
   count = 1;

  }
 }

 public int size(){
return size(root);

 }
public void c(){

    System.out.println(root.count);
}

public void put(K key,V value){

root=put(root, key, value);
}

public Object get(K key){
    return contains(root, key);
}

public void delete(K key){
  
    root=delete(root, key);

}

public Object r(){
    return root.rightChild.key;
}

public int rank(K key){

    return rank(root,key);
}

public Iterable<K> keys(){
Queue<K> q=new LinkedList<K>() ;
inorder(root,q);

return q;

}
public K max(){return max(root);
}
public Object min(){
return min(root).key;

}
/* ============================== */
private Object contains(Node node,K key){
    if(node==null){
return null;


    }
int cmp=((Comparable <K>)key).compareTo((K)node.key);

    if(cmp==0)
    return node.value;

if(cmp>0){
    return contains(node.rightChild, key);
}
else 
return contains(node.leftChild, key);

}

private void inorder(Node node,Queue q){
if(node==null) return;
inorder(node.leftChild, q);
q.add(node.key);
inorder(node.rightChild, q);

}
private Node deleteMin(Node node){
if(node.leftChild==null) return node.rightChild;
node.leftChild=deleteMin(node.leftChild);
node.count=1+size(node.leftChild)+size(node.rightChild);
return node;

}

private Node delete(Node node,K key){
if(node==null){
return null;
}
int cmp=((Comparable <K>)(key)).compareTo((K)node.key);
if(cmp>0){return node.leftChild=delete(node.leftChild, key);}
if (cmp<0){return node.rightChild=delete(node.rightChild, key);}

/* if we find the node.key==key ie cmp==0 */
else
{                                                                                          
    /* if node has only one child either left or right we return according to it*/

    if(node.rightChild==null){return node.leftChild;} 
 if(node.leftChild==null){return node.rightChild;}

 /* otherwise if both the chidren are present*/
Node x=node;
node=(Node) min(x.rightChild);
node.rightChild=deleteMin(x.rightChild);
node.leftChild=x.leftChild;
}
node.count=1+size(node.leftChild)+size(node.rightChild);

return node;
}

private int rank(Node node,K key)
{
if (node==null) return 0;
int cmp=((Comparable <K>)(key)).compareTo((K)node.key);

if(cmp<0)
    return rank(node.leftChild,key);

    else if(cmp>0)

    return 1+size(node.leftChild)+rank(node.rightChild,key);
else{
return size(node.leftChild);
}
}

private K max(Node node){
    if(node.rightChild==null){
        return (K)node.key;
    }
    return max(node.rightChild);
}

 
private Node min(Node node){
if(node.leftChild==null)
return node;
else
return min(node.leftChild);


}
private Node put(Node<K,V> node,K key,V value){
 if(node==null){ 
 return  new Node(key,value);
}
 int cmp=((Comparable <K>)key).compareTo(node.key);
 if(cmp>0){
 node.rightChild=put(node.rightChild, key, value);
 }
 else
{ node.leftChild=put(node.leftChild, key, value);
}

node.count=1+size(node.leftChild)+size(node.rightChild);
return node;
}

private int size(Node node){
if(node==null){
return 0;}
return node.count;

}


}